base/shell/explorer/traydeskbtn.cpp

```diff
@@ -0,0 +1,443 @@
+}
+
+LRESULT CTrayShowDesktopButton::OnDestroy(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    if (m_hTheme)
+    {
+        ::CloseThemeData(m_hTheme);
+        m_hTheme = NULL;
+    }
+    if (m_hFallbackTheme)
+    {
+        ::CloseThemeData(m_hFallbackTheme);
+        m_hFallbackTheme = NULL;
+    }
+
+    return 0;
+}
+
+VOID CTrayShowDesktopButton::OnDraw(HDC hdc, LPRECT prc)
+{
+    RECT rc = { prc->left, prc->top, prc->right, prc->bottom };
+    LPRECT lpRc = &rc;
+    HBRUSH hbrBackground = NULL;
+
+    if (m_hTheme)
+    {
+        HTHEME theme;
+        int part = 0;
+        int state = 0;
+        if (m_drawWithDedicatedBackground)
+        {
+            theme = m_hTheme;
+
+            if (m_bPressed)
+                state = PBS_PRESSED;
+            else if (m_bHovering)
+                state = PBS_HOT;
+            else
+                state = PBS_NORMAL;
+        }
+        else
+        {
+            part = TP_BUTTON;
+            theme = m_hFallbackTheme;
+
+            if (m_bPressed)
+                state = TS_PRESSED;
+            else if (m_bHovering)
+                state = TS_HOT;
+            else
+                state = TS_NORMAL;
+
+            if (IsHorizontal)
+                rc.right -= m_inset.cx;
+            else
+                rc.bottom -= m_inset.cy;
+        }
+
+        if (::IsThemeBackgroundPartiallyTransparent(theme, part, state))
+            ::DrawThemeParentBackground(m_hWnd, hdc, NULL);
+
+        ::DrawThemeBackground(theme, hdc, part, state, lpRc, lpRc);
+    }
+    else
+    {
+        hbrBackground = ::GetSysColorBrush(COLOR_3DFACE);
+        ::FillRect(hdc, lpRc, hbrBackground);
+
+        if (m_bPressed || m_bHovering)
+        {
+            UINT edge = m_bPressed ? BDR_SUNKENOUTER : BDR_RAISEDINNER;
+            DrawEdge(hdc, lpRc, edge, BF_RECT);
+        }
+    }
+
+    if (m_highContrastMode || !m_drawWithDedicatedBackground)
+    {
+        /* Prepare to draw icon */
+
+        // Determine X-position of icon's top-left corner
+        int iconX = rc.left;
+        iconX += (rc.right - iconX) / 2;
+        iconX -= m_szIcon.cx / 2;
+
+        // Determine Y-position of icon's top-left corner
+        int iconY = rc.top;
+        iconY += (rc.bottom - iconY) / 2;
+        iconY -= m_szIcon.cy / 2;
+
+        // Ok now actually draw the icon itself
+        if (m_icon)
+        {
+            DrawIconEx(hdc, iconX, iconY,
+                m_icon, 0, 0,
+                0, hbrBackground, DI_NORMAL);
+        }
+        else // Fallback for if icon isn't available or something idk lol
+        {
+            RECT rcIconFallback = { iconX, iconY, iconX + m_szIcon.cx, iconY + m_szIcon.cy };
+            HBRUSH hbrFallback;
+
+            // Border - #808080
+            hbrFallback = CreateSolidBrush(RGB(0x80, 0x80, 0x80));
+            ::FillRect(hdc, &rcIconFallback, hbrFallback);
+            ::DeleteObject(hbrFallback);
+
+            // Background - #3A6EA5
```

Comment:
Aren't these values theme-specific?

Reply:
No

Reply:
So, what happens if I change the color scheme of ReactOS (for example in the classic scheme)?

Reply:
@HBelusca Not relevant - those colours are for a fallback case where I programmatically draw an approximation of an icon, in the event that the actual icon resource somehow can't be retrieved.

I originally implemented this fallback behaviour back when I'd still retrieved the icon from `imageres.dll`, as Windows 7 does, for sake of compatibility with/cross-testing on versions of Windows which lack `imageres.dll` and thus lack the relevant icon - for example, Windows Server 2003 - but later on, you folks directed me to pull the icon from `shell32.dll`, since it's a suitable duplicate on Windows 7+ and also exists even as far back as Windows XP.

With all that in mind...this fallback behaviour's sole (known) purpose for existing has been rendered 100% moot, and now it's causing more and more review confusion, so...should I perhaps remove the fallback behaviour entirely? I kind of think so...what do you think?

Reply:
If the icon cannot be loaded for some reason, you usually just don't render the icon. Having a colored square icon in place of it may cause confusion, making people think is a really strange bug. There is no need for a fallback in this case.

Reply:
ok will do